feat: Displaying pre-selected single namespace

anshu-rani-singh opened the pull request:
## Checklist

- [x] Follow the [contributor guide](./CONTRIBUTING.md)
- [ ] End-to-end tests pass
- [ ] Unit tests for essential parts of your code, e.g. APIs exposed via SDKs or endpoints
- [ ] Update `./docs`
- [ ] Recognize contributors with "@all-contributors please add <person> for <code, bug, docs, etc>"
- [x] Linked issues #eg101 

Jira: https://mckinsey.atlassian.net/browse/ARKQB-229

anshu-rani-singh committed:
Showing current namespace

anshu-rani-singh committed:
Merge branch 'main' of github.com:mckinsey/agents-at-scale-ark into feat/pre-selected-namespace

anshu-rani-singh committed:
Merge branch 'main' of github.com:mckinsey/agents-at-scale-ark into feat/pre-selected-namespace

dwmkerr commented on .ark.env.local:
```diff
@@ -4,4 +4,5 @@ ARK_QUICKSTART_BASE_URL=https://azure.prod.com
-ARK_QUICKSTART_CONTROLLER_IMAGE=
\ No newline at end of file
+ARK_QUICKSTART_CONTROLLER_IMAGE=
+ARK_DETECT_NAMESPACE=
```
I think we can remove this env var completely - will discuss later in the PR to show how

dwmkerr commented on services/ark-api/ark-api/src/ark_api/core/namespace.py:
```diff
@@ -0,0 +1,55 @@
+_current_namespace: str = "default"
+
+
+def is_namespace_detection_enabled() -> bool:
```
OK this is looking great - I think we can actually remove this function and instead do the following:

In the ark-sdk we have two mechanisms to get resources such as queries, the first is that we just try and call the kubernetes apis (which works if we are running the ark-apis in cluster) or instead we read the `~/.kube/config` file. This means in the `ark-sdk` we can have a helper function that uses its state to get the current kube context, we could have an sdk function like `get_context()` which could return this:

```
{
    "namespace": "default",
    "cluster": "ark-cluster"
  }
```

Only `namespace` is important, the other fields are nice to have. this should work pretty much out of the box - if it doesn't then the ark-sdk would return the kube config context and then in the api layer if it doesn't know the context then we can use the `/var` mount (which should be preferred).

Then in the ark-apis, we could have something very similar to `kubectl get-context 

```
GET /v1/context
{
    "namespace": "default",
    "cluster": "ark-cluster"
  }
```

We'd call this get context api from the dashboard just like you suggested - and the namespace selector box would still be a selector box with just one entry - the current namespace.

The common pattern in k8s tools is this:

```
// Read namespace once at startup
  namespace, err := ioutil.ReadFile("/var/run/secrets/kubernetes.io/serviceaccount/namespace")
  if err != nil {
      namespace = "default"  // fallback
  }
```

We'd use ark-sdk instead, which reads the `/var` file and if not set falls back to ~/.kube and if not set falls back to `default` - that'd follow the patterns of tools like `helm` etc

notes for where to find the kube context stuff in ark-sdk:

sdk/gen_sdk/overlay/python/ark_sdk/k8s.py
- First try to load kubeconfig from the default location (~/.kube/config) - this uses the kubectl context (line 33)
- If that fails, try in-cluster config - this is for when running inside a Kubernetes pod (line 44)

dwmkerr commented on services/ark-api/ark-api/src/ark_api/core/namespace.py:
```diff
@@ -0,0 +1,55 @@
+        logger.info("Namespace detection disabled via ARK_DETECT_NAMESPACE, using default")
+        return "default"
+
+    namespace_file = Path("/var/run/secrets/kubernetes.io/serviceaccount/namespace")
```
as per above this can probably move into `ark-sdk` so that it uses this file, falls back to `.kube` context and falls back to default

dwmkerr commented on services/ark-dashboard/ark-dashboard/components/app-sidebar.tsx:
```diff
@@ -55,23 +56,71 @@ export function AppSidebar() {
-        if (urlNamespace && filteredNamespaces.some(ns => ns.name === urlNamespace)) {
-          setNamespace(urlNamespace)
+        // Try to get current namespace and mode information
+        let selectedNamespace = "default"
```
once we've made the described changes on startup we can just set the url param once to the current context namespace and it'll then simply continue to use that, no need for single/multi mode, we just only show one namespace in the drop down box (then later for cluster admins we can bring back the other namespaces)

dwmkerr commented on services/ark-dashboard/ark-dashboard/lib/api/generated/types.ts:
```diff
@@ -82,6 +82,29 @@ export interface paths {
         patch?: never;
         trace?: never;
     };
+    "/v1/current-namespace": {
```
the this becomes `v1/context`

dwmkerr reviewed (changes requested):
Looks great a few suggestions!

anshu-rani-singh committed:
Addressed review comments

anshu-rani-singh committed:
Resolved conflict

anshu-rani-singh committed:
Updating type as per new api

dwmkerr committed:
Merge branch 'main' into feat/pre-selected-namespace

dwmkerr committed:
feat: simplify namespace detection and remove env variable

dwmkerr committed:
fix: improve namespace context detection in ark-sdk (#125)

## Summary
- Fix namespace detection bugs (caching, cluster field, default
override)
- Remove ARK_DETECT_NAMESPACE environment variable
- Simplify dashboard to use only detected namespace (no list API call)

anshu-rani-singh committed:
Merge branch 'main' of github.com:mckinsey/agents-at-scale-ark into feat/pre-selected-namespace
